[Serve] [Docs] Document best practices for querying Serve over HTTP (#37160)
ray-project/ray: doc/source/serve/production-guide/best-practices.md

(serve-best-practices)=

# Best practices in production

> This section will help you:

Reviewer: ```suggestion
This section helps you:
```

> * Understand best practice when operating Serve in production

Reviewer: ```suggestion
* Understand best practices when operating Serve in production
```

* Learn more about managing Serve with the Serve CLI
* Configure your HTTP requests when querying Serve

## CLI best practices

> This section summarizes the best practices when deploying to production using the Serve CLI:

Reviewer: ```suggestion
This section summarizes the best practices for deploying to production using the Serve CLI:
```

  message: ''
```

For Kubernetes deployments with KubeRay, tighter integrations of `serve status` with Kubernetes are available. See [Getting the status of Serve applications in Kubernetes](serve-getting-status-kubernetes).

## Best practices for HTTP requests

Most examples in these docs use straightforward `get` or `post` requests using Python's `requests` library, such as:

```{literalinclude} ../doc_code/requests_best_practices.py
:start-after: __prototype_code_start__
:end-before: __prototype_code_end__
:language: python
```

This pattern is useful for prototyping, but it isn't sufficient for production. In production, HTTP requests should use:

> * Retries: requests may occasionally fail due to transient issues (e.g. slow network, node failure, power outage, spike in traffic, etc.). Retry failed requests a handful (~10) of times to account for these issues.

Reviewer: ```suggestion
* Retries: Requests may occasionally fail due to transient issues (e.g., slow network, node failure, power outage, spike in traffic, etc.) Retry failed requests about 10 times to account for these issues.
```

Author: I'd rather remove 10 here, since I'm not sure if that's the ideal number across different use cases. Let me make the change.

> * Exponential backoff: To avoid bombarding the Serve application with retries during a transient error, apply an exponential backoff on failure. Each retry should wait exponentially longer than the previous one before running. For example, the first retry may happen 0.1 seconds after a failure, the next one 0.4s (4 x 0.1), then 1.6s, 6.4s, 25.6s, etc.

Reviewer: ```suggestion
* Exponential backoff: To avoid bombarding the Serve application with retries during a transient error, apply an exponential backoff on failure. Each retry should wait exponentially longer than the previous one before running. For example, the first retry may happen 0.1s after a failure, and subsequent retries happen 0.4s (4 x 0.1), 1.6s, 6.4s, 25.6s, etc. after the failure.
```

> * Timeouts: add a timeout to each retry to prevent your requests from hanging. The timeout should be longer than the your application's latency to give your application enough time to process requests. Additionally, we recommend setting an [end-to-end timeout](serve-performance-e2e-timeout) in the Serve application, so slow requests don't bottleneck replicas.

Reviewer: ```suggestion
* Timeouts: Add a timeout to each retry to prevent requests from hanging. The timeout should be longer than the application's latency to give your application enough time to process requests. Additionally, set an [end-to-end timeout](serve-performance-e2e-timeout) in the Serve application, so slow requests don't bottleneck replicas.
```
